### PR TITLE
Add credentialScope to endpoint params

### PR DESCRIPTION
**Pull request by davidh44**
## Modifications

Update codegen to add `credentialScope` to endpoint params
Add credentialScope `BuiltInParameter`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by joviegas** [comment]
The Specs says

"This MUST only be used by global services"

How are we ensuring this ?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java [hunk 1]
```diff
@@ -98,12 +100,27 @@ public static QueryEndpointParams ruleParams(SdkRequest request, ExecutionAttrib
...
         setClientContextParams(builder, executionAttributes);
         setContextParams(builder, executionAttributes.getAttribute(AwsExecutionAttribute.OPERATION_NAME), request);
         setStaticContextParams(builder, executionAttributes.getAttribute(AwsExecutionAttribute.OPERATION_NAME));
         return builder.build();
     }
 
+    private static Region getCredentialScopeIfPresent(ExecutionAttributes executionAttributes) {
+        SelectedAuthScheme<?> authScheme = executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
+        try {
+            Identity identity = authScheme.identity().get();
```
Comment (joviegas) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java:113]: Can we avoid this blocking call?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/auth/scheme/query-endpoint-auth-params-auth-scheme-params-without-allowlist.java [hunk 2]
```diff
@@ -1,18 +1,3 @@
-/*
```
[the remark this answers is not in the document]
Comment (davidh44) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/auth/scheme/query-endpoint-auth-params-auth-scheme-params-without-allowlist.java:1]: added back

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/auth/scheme/query-endpoint-auth-params-without-allowlist-auth-scheme-interceptor.java [hunk 3]
```diff
@@ -102,8 +102,7 @@ private QueryAuthSchemeParams authSchemeParams(SdkRequest request, ExecutionAttr
     }
 
     private <T extends Identity> SelectedAuthScheme<T> trySelectAuthScheme(AuthSchemeOption authOption, AuthScheme<T> authScheme,
-                                                                           IdentityProviders identityProviders, List<Supplier<String>> discardedReasons,
-                                                                           MetricCollector metricCollector) {
+                                                                           IdentityProviders identityProviders, List<Supplier<String>> discardedReasons, MetricCollector metricCollector) {
```
[the remark this answers is not in the document]
Comment (davidh44) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/auth/scheme/query-endpoint-auth-params-without-allowlist-auth-scheme-interceptor.java]: reverted

**Comment by davidh44** [comment]
> The Specs says
> 
> "This MUST only be used by global services"
> 
> How are we ensuring this ?

This will be enforced by rulesets

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java (hunk 1 above)
Comment (davidh44) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java]: discussed offline, will use joinLikeSync()

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java [hunk 4]
```diff
@@ -98,12 +101,27 @@ public static QueryEndpointParams ruleParams(SdkRequest request, ExecutionAttrib
...
+    private static Region getCredentialScopeIfPresent(ExecutionAttributes executionAttributes) {
+        SelectedAuthScheme<?> authScheme = executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
+        try {
+            Identity identity = CompletableFutureUtils.joinLikeSync(authScheme.identity());
+            if (identity instanceof AwsCredentialsIdentity) {
+                AwsCredentialsIdentity awsCredentialsIdentity = (AwsCredentialsIdentity) identity;
+                return awsCredentialsIdentity.credentialScope().map(Region::of).orElse(null);
+            }
+        } catch (Exception error) {
+            throw SdkClientException.create(error.getMessage());
```
Comment (zoewangg) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java:120]: Why are we dropping the cause?

Comment (davidh44) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/rules/endpoint-resolve-interceptor-preSra.java]: added cause

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/EndpointResolverInterceptorSpec.java [hunk 5]
```diff
@@ -295,6 +306,33 @@ private MethodSpec ruleParams() {
...
+
+        b.addStatement("$T identity = $T.joinLikeSync(authScheme.identity())", Identity.class, CompletableFutureUtils.class);
+        b.beginControlFlow("if (identity instanceof $T)", AwsCredentialsIdentity.class);
+        b.addStatement("$T awsCredentialsIdentity = ($T) identity", AwsCredentialsIdentity.class, AwsCredentialsIdentity.class);
+        b.addStatement("return awsCredentialsIdentity.credentialScope().map($T::of).orElse(null)", Region.class);
+        b.endControlFlow();
+
+        b.endControlFlow();
+
+        b.beginControlFlow("catch ($T error)", Exception.class);
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/EndpointResolverInterceptorSpec.java:327]: I dont think we need try catch here.

The [joinLikeSync](https://github.com/aws/aws-sdk-java-v2/blob/becd54c88a5ac611957bad242d2823971ac5bd63/utils/src/main/java/software/amazon/awssdk/utils/CompletableFutureUtils.java#L247-L259) already throws a Runtime exception

**Review by joviegas — APPROVED** [review]

**Pull request merged**